gcc/rust/checks/errors/borrowck/rust-bir.h

```diff
@@ -0,0 +1,200 @@
+// Copyright (C) 2020-2023 Free Software Foundation, Inc.
+
+// This file is part of GCC.
+
+// GCC is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3, or (at your option) any later
+// version.
+
+// GCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with GCC; see the file COPYING3.  If not see
+// <http://www.gnu.org/licenses/>.
+
+#ifndef RUST_BIR_BASE_H
+#define RUST_BIR_BASE_H
+
+#include "rust-bir-place.h"
+#include "rust-bir-visitor.h"
+
+namespace Rust {
+
+namespace BIR {
+
+struct BasicBlock;
+class Node;
+class AbstractExpr;
+
+/**
+ * Top-level entity of the Borrow-checker IR (BIR).
+ * It represents a single function (method, closure, etc.), which is also the
+ * basic unit of Polonius borrow-checking.
+ */
+struct Function
+{
+  PlaceDB place_db;
+  std::vector<PlaceId> arguments; // Only used for dump.
+  std::vector<BasicBlock> basic_blocks;
+};
+
+/** Single "instruction" of BIR. */
+class Node
+{
+public:
+  enum Kind
+  {
+    ASSIGNMENT,	  // <place> = <expr>
+    SWITCH,	  // switch <place>
+    RETURN,	  // return
+    GOTO,	  // goto
+    STORAGE_DEAD, // StorageDead(<place>)
+    STORAGE_LIVE, // StorageLive(<place>)
+  };
+
+private:
+  Kind kind;
+  // ASSIGNMENT: lhs
+  // SWITCH: switch_val
+  // StorageDead/StorageLive: place
+  // otherwise: <unused>
+  PlaceId place;
+  // ASSIGNMENT: rhs
+  // otherwise: <unused>
+  std::unique_ptr<AbstractExpr> expr;
```

Comment:
so in Rust this would be something like this, correct?

```rust
enum Kind {
    Assignment { lhs: PlaceId, rhs: AbstractExpr },
    Switch { switch_val: PlaceId },
    StorageDead(PlaceId),
    StorageAlive(PlaceId),
    Goto,
    Return,
}
```

if that is the case, then I think a better API would be to have methods like `get_lhs()`, `get_rhs()` or `get_switch_value()` which would `rust_assert()` that we have the proper kind. does that make sense?

Reply:
>so in Rust this would be something like this, correct?

Yes

> if that is the case, then I think a better API would be to have methods like get_lhs(), get_rhs() or get_switch_value() which would rust_assert() that we have the proper kind. does that make sense?

Sure, sounds good. I am not convinced it is actually useful here, but it might as the checks get more complicated.